certstore_linux.go

```diff
@@ -1,14 +1,369 @@
 package certstore
 
-import "errors"
+/*
+#cgo CFLAGS: -I/usr/include/nss -I/usr/include/nspr
+#cgo LDFLAGS: -lnss3 -lnspr4 -lsmime3
+#include <nss.h>
+#include <pk11pub.h>
+#include <nspr.h>
+#include <prio.h>
+#include <prerror.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+#include <p12.h>
+#include <p12plcy.h>
+#include <stdlib.h>
+#include <cryptohi.h>
 
-// This will hopefully give a compiler error that will hint at the fact that
-// this package isn't designed to work on Linux.
-func init() {
-	CERTSTORE_DOESNT_WORK_ON_LINIX
+SECItem *P12U_NicknameCollisionCallback(SECItem *old_nick, PRBool *cancel, void *wincx) {
+	char *nick = NULL;
+	SECItem *ret_nick = NULL;
+	CERTCertificate *cert = (CERTCertificate *)wincx;
+	if (!cancel || !cert) {
+		fprintf(stdout, "cancel: %p, cert: %p\n", cancel, cert);
+		return NULL;
+	}
+	nick = CERT_MakeCANickname(cert);
+	if (!nick) {
+		fprintf(stdout, "nick %p\n", nick);
+		return NULL;
+	}
+	if (old_nick && old_nick->data && old_nick->len &&
+		PORT_Strlen(nick) == old_nick->len &&
+		!PORT_Strncmp((char *)old_nick->data, nick, old_nick->len)) {
+		PORT_Free(nick);
+		fprintf(stdout, "old_nick %p, nick %p\n", old_nick, nick);
+		return NULL;
+    }
+	//fprintf(stdout, "using nickname: %s\n", nick);
+	ret_nick = PORT_ZNew(SECItem);
+	if (ret_nick == NULL) {
+		PORT_Free(nick);
+		fprintf(stdout, "ret_nick %p\n", ret_nick);
+		return NULL;
+	}
+	ret_nick->data = (unsigned char *)nick;
+	ret_nick->len = PORT_Strlen(nick);
+	return ret_nick;
+}
+*/
+import "C"
+import (
+	"crypto"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"syscall"
+	"unicode/utf16"
+	"unsafe"
+)
+
+type Passwd struct {
+	PwName   string
+	PwPasswd string
+	PwUid    int
+	PwGid    int
+	PwGecos  string
+	PwDir    string
+	PwShell  string
+}
+type nssStore int
+
+func (nssStore) Identities() ([]Identity, error) {
+	var identities = make([]Identity, 0)
+	//fmt.Printf("Listing certificates:\n")
+	var certs = C.PK11_ListCerts(C.PK11CertListType(C.PK11CertListAll), unsafe.Pointer(nil))
```

Comment:
Does this only find certs via PKCS11, or does it return locally stored certificates also?

Reply:
yes, all locally stored certs (p12 imported in google chrome e.g. are visible by this call)

Reply:
Great!